### PR TITLE
Make: enable -Werror -Wall -Wextra by default

### PR DESCRIPTION
**Pull request by smlng**
This picks up the idea of #1121 (which isn't mergeable anymore, because of considerable changes since then) to enable more compiler warnings and make them errors to fail. This PR also ships several fixes to make things work.

It also provides an option to turn on `-Wpedantic` too, however, that would require even more work and also some agreement (e.g., usage of gcc extensions, see #7918).

may need #7860 and other existing PR that solve open issues.

### CODE REVIEW

**Comment by smlng** [comment]
also tracked issues in #7875 need to be fixed for this.

**Comment by kaspar030** [comment]
While most of the fix commits are trivial, some are not, which will make reviewing difficult.

I suggest you make this PR only enable the extra warnings, and then open smaller PR's fixing the actual compile issues. You could group them (e.g., all "unused parameter" fixes, then all "signed compare" ...).

**Comment by smlng** [comment]
@kaspar030 you're right, on first sight I thought: just a few fixes, make it one PR. But then the list got longer and longer and longer ... as always, will split in groups as suggested.

File: cpu/cc2538/include/cpu_conf.h [hunk 1]
```diff
@@ -20,6 +20,9 @@
 #define CPU_CONF_H
 
 #include "cpu_conf_common.h"
+/* store compiler switches */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
```
Comment (cladmi) [inline cpu/cc2538/include/cpu_conf.h]: For this warning ignoring, I would prefer to see why its disabled than "store compiler switches".

File: cpu/cc26x0/include/cpu_conf.h [hunk 2]
```diff
@@ -22,8 +22,10 @@
 
 #include "cpu_conf_common.h"
 
+/* store compiler switches */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
```
Comment (cladmi) [inline cpu/cc26x0/include/cpu_conf.h]: Same here, comment why pedantic is disabled.

File: cpu/k22f/include/cpu_conf.h [hunk 3]
```diff
@@ -22,11 +22,16 @@
 #ifndef CPU_CONF_H
 #define CPU_CONF_H
 
+/* store compiler switches */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
```
Comment (cladmi) [inline cpu/k22f/include/cpu_conf.h]: Same here, comment why pedantic is disabled.

File: cpu/k60/include/cpu_conf.h [hunk 4]
```diff
@@ -22,6 +22,9 @@
 #ifndef CPU_CONF_H
 #define CPU_CONF_H
 
+/* store compiler switches */
```
Comment (cladmi) [inline cpu/k60/include/cpu_conf.h]: Same here, comment why pedantic is disabled.

File: cpu/k64f/include/cpu_conf.h [hunk 5]
```diff
@@ -24,11 +24,16 @@
 #ifndef CPU_CONF_H
 #define CPU_CONF_H
 
+/* store compiler switches */
```
Comment (cladmi) [inline cpu/k64f/include/cpu_conf.h]: Same here, comment why pedantic is disabled.

File: cpu/kw2xd/include/cpu_conf.h [hunk 6]
```diff
@@ -29,6 +29,9 @@
 #ifndef CPU_CONF_H
 #define CPU_CONF_H
 
+/* store compiler switches */
```
Comment (cladmi) [inline cpu/kw2xd/include/cpu_conf.h]: Same here, comment why pedantic is disabled.

File: cpu/saml21/periph/timer.c [hunk 7]
```diff
@@ -48,6 +48,10 @@ static inline void _irq_enable(tim_t dev);
  */
 int timer_init(tim_t dev, unsigned long freq, timer_cb_t cb, void *arg)
 {
+    /* at the moment, the timer can only run at 1MHz */
```
Comment (cladmi) [inline cpu/saml21/periph/timer.c]: This is not related to -Wall -Wextra, or is it ?

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 8]
```diff
@@ -138,8 +138,10 @@ uint8_t at86rf2xx_get_chan(at86rf2xx_t *dev)
 
 void at86rf2xx_set_chan(at86rf2xx_t *dev, uint8_t channel)
 {
-    if ((channel < AT86RF2XX_MIN_CHANNEL) ||
-        (channel > AT86RF2XX_MAX_CHANNEL) ||
+    if ((channel > AT86RF2XX_MAX_CHANNEL) ||
+#if AT86RF2XX_MIN_CHANNEL /* is zero for sub-GHz */
```
Comment (cladmi) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: Maybe duplicate the if lines for readability here, like in _netdev.c

File: cpu/cc2538/include/cpu_conf.h (hunk 1 above)
Comment (smlng) [reply cpu/cc2538/include/cpu_conf.h]: yes right, here and below its mostly to exclude vendor headers from being checked with pedantic on - because we don't want to change or fix them but rather use the originals and also update them from time to time

**Comment by smlng** [comment]
separated into fixes into groups as suggested by @kaspar030, this PR now requires #7993, #7994, #7995, and other before merge.

**Comment by smlng** [comment]
rebased onto #7993, #7994, #7995 -> they are now included here, to check with Murdock whats still missing,

**Comment by cladmi** [comment]
Please let us know when you finished the sub-PRs so we can review them.

**Comment by smlng** [comment]
@cladmi yep, will do - forgot to mark it WIP, sorry

**Comment by smlng** [comment]
@kaspar030, @gebart, @miri64 this one finally got through w/o errors, if we act swiftly by getting #7993, #7994, and #7995 merged, we can have `-Wall -Werror -Wextra` activated - before something/one breaks it again 😉 . And maybe work on `-Wpedantic`

**Comment by miri64** [comment]
> @kaspar030, @gebart, @miri64 this one finally got through w/o errors, if we act swiftly by getting #7993, #7994, and #7995 merged, we can have -Wall -Werror -Wextra activated - before something/one breaks it again  . And maybe work on -Wpedantic

It's not possible to merge https://github.com/RIOT-OS/RIOT/pull/7994 with the GNRC embargo up.

**Comment by miri64** [comment]
As far as I understand the PRs are devided up by error category. Maybe we can take out the error category #7994 is address in for now and reintroduce it after the GNRC embargo us lifted.

**Comment by miri64** [comment]
GNRC embargo is lifted => needs rebase

**Comment by smlng** [comment]
rebased after merge of #7994  and #7995, now excluded #7993 which is only needed for `-Wpedantic`

**Comment by miri64** [comment]
Seems like some new errors were introduced in #7925 ^^"

**Comment by smlng** [comment]
yeah will add fixes right away, please hold the line 😉

**Comment by miri64** [comment]
Travis issue is the same (now proven wrong in #8164) issue as in #7994.

**Comment by smlng** [comment]
> Travis issue is the same (now proven wrong in #8164) issue as in #7994.

yeah, saw that too - can fix it somehow anyway - maybe some cppcheck suppression stuff or so?

**Comment by miri64** [comment]
> yeah, saw that too - can fix it somehow anyway - maybe some cppcheck suppression stuff or so?

If you are hell-bend on having a green tick on your PR you can do that (with `/* cppcheck-suppress duplicateBranch */` (+ reasoning, see https://github.com/RIOT-OS/RIOT/pull/6024)... But since this *should* only be pointed out here, because you are changing that file, we also could do that in a separate PR. The green tick by Murdock is the one that counts.

**Comment by smlng** [comment]
>  we also could do that in a separate PR.

sure, though I like all to have all green 😉

**Comment by miri64** [comment]
When this get's merged today, you should write a reminder to the maintainers (and maybe even devel) mailing list that all builds on Murdock should be rebuild before merging ;-). Otherwise older PRs could re-introduce new bugs.

**Comment by smlng** [comment]
murdock looks good, may I squash?

**Comment by miri64** [comment]
Yes, please!

**Comment by smlng** [comment]
squashed

**Comment by smlng** [comment]
@cladmi can you have another look, so we can merge this soonish! Merci!

**Comment by smlng** [comment]
ping! This might be worth two (or more) ACKs before merge?!

File: cpu/saml21/periph/timer.c (hunk 7 above)
Comment (miri64) [reply cpu/saml21/periph/timer.c]: > This is not related to -Wall -Wextra, or is it ?

Ping @smlng?

File: drivers/at86rf2xx/at86rf2xx_getset.c (hunk 8 above)
Comment (miri64) [reply drivers/at86rf2xx/at86rf2xx_getset.c]: > Maybe duplicate the if lines for readability here, like in _netdev.c

Ping @smlng?

Comment (smlng) [reply drivers/at86rf2xx/at86rf2xx_getset.c]: well, maybe not?

File: cpu/saml21/periph/timer.c (hunk 7 above)
Comment (smlng) [reply cpu/saml21/periph/timer.c]: mhm, IIRC copy pasted from `samd21/periph/timer.c`, but don't know why/how it slipped in here. Shall I remove?

Comment (cladmi) [reply cpu/saml21/periph/timer.c]: I don't care it can stay.

File: drivers/at86rf2xx/at86rf2xx_getset.c (hunk 8 above)
Comment (cladmi) [reply drivers/at86rf2xx/at86rf2xx_getset.c]: Same here, its just style.

**Comment by miri64** [comment]
(The rebuild rule should be applied here as well, before hitting the green button, btw ;-))

File: tests/trickle/main.c [hunk 9]
```diff
@@ -53,14 +53,16 @@ static void callback(void *args)
     return;
 }
 
-static trickle_t trickle = { .callback.func = &callback,
-                             .callback.args = NULL };
+static trickle_t trickle;
```
Comment (cladmi) [inline tests/trickle/main.c]: What was the problem with this one ?

Comment (smlng) [reply tests/trickle/main.c]: the way of init the struct caused an error.

Comment (miri64) [reply tests/trickle/main.c]: maybe use

```C
static trickle_t trickle = { .callback = { .func = &callback,
                                           .args = NULL } };
```

instead? It doesn't change the behavior and is the more correct version.

File: drivers/nrf24l01p/nrf24l01p.c [hunk 10]
```diff
@@ -353,7 +354,7 @@ int nrf24l01p_set_payload_width(const nrf24l01p_t *dev, nrf24l01p_rx_pipe_t pipe
             return -1;
     }
 
-    if (width < 0) {
+    if (width == 0) {
```
[the remark this answers is not in the document]
Comment (miri64) [reply drivers/nrf24l01p/nrf24l01p.c]: No... `width < 0` is not equivalent to `width == 0`, even if unsigned... An unsigned can be 0, and `width < 0` would still be false.

File: tests/trickle/main.c (hunk 9 above)
Comment (cladmi) [reply tests/trickle/main.c]: For native, iotlab-m3, samr21-xpro I get no error with the first version.

What was the error ? the commit message does not help me "tests/trickle: fix struct init"

File: drivers/nrf24l01p/nrf24l01p.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (smlng) [reply drivers/nrf24l01p/nrf24l01p.c]: you're right, its not the same. However, width is unsigned so `< 0` makes no sense, further it also makes no sense to set payload length `0`, too. So the check is still valid (IMHO).

File: tests/trickle/main.c (hunk 9 above)
Comment (smlng) [reply tests/trickle/main.c]: Error was reported by `clang` on macOS, which is often more pedantic than `gcc`

File: drivers/nrf24l01p/nrf24l01p.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (miri64) [reply drivers/nrf24l01p/nrf24l01p.c]: But it changes the meaning completely now (let's not even talk about the API change). If 0 wasn't an error case before, you made it one now.

[the remark this answers is not in the document]
Comment (smlng) [reply drivers/nrf24l01p/nrf24l01p.c]: well then the checked can be removed completely, if that's the (intermediate) solution - leaving a side that `width = 0` should be wrong anyway.

File: tests/trickle/main.c (hunk 9 above)
Comment (smlng) [reply tests/trickle/main.c]: mhm, can't reproduce right now - maybe it was reported by Murdock on some other platform (too). At least the old init version didn't work, @miri64 solution looks okay to me, too.

File: drivers/nrf24l01p/nrf24l01p.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (miri64) [reply drivers/nrf24l01p/nrf24l01p.c]: Since this PR is not about changing any functional errors I would prefer the intermediate solution and a fix as a follow-up.

[the remark this answers is not in the document]
Comment (smlng) [reply drivers/nrf24l01p/nrf24l01p.c]: okay

**Comment by cladmi** [comment]
Only need to squash and wait for murdock.

**Comment by smlng** [comment]
squashed
